[5.3] Refactor away from call_user_func[_array] in stack sensitive areas

ralphschindler opened the pull request:
Instead of calling a function (`call_user_func` & `call_user_func_array`), call the target callable directly.  This has the benefit of tightening up the stack trace making it easier to read through. At current, introducing a middleware to the pipeline requires 5 function calls to dispatch and unravel the remaining pipelined middlewares.  This patch reduces the number of function calls required to 3 per middleware.  This could potentially have the side-effect of being a micro-optimization, but that is not the intent.

As an example, throw an exception from the Hello World controller.  At this point in the default dispatch cycle, the controller is 55 frames deep.  After this change, we've reduced the stack from 55 to 36 frames.  Essentially, every new Middleware that is introduced before the RouteServiceProvider/dispatcher will require 5 function calls currently, and only 3 after. Images attached...

Before:
![framework-stacktrace-small](https://cloud.githubusercontent.com/assets/76674/17459069/8b4a6c0c-5bf0-11e6-866f-cecd4844a733.jpg)

After:
![framework-stacktrace-after-small](https://cloud.githubusercontent.com/assets/76674/17459072/9f2446bc-5bf0-11e6-8bee-0f1e586b3ac2.jpg)

ralphschindler committed:
Refactor away from call_user_func[_array]
- in Pipeline and Routing/Router

GrahamCampbell commented on src/Illuminate/Pipeline/Pipeline.php:
```diff
@@ -134,7 +133,14 @@ protected function getSlice()
                 }
 
-                return call_user_func_array([$pipe, $this->method], $parameters);
+                switch (count($parameters)) {
```
This isn't more efficient. Better just to use unpacking.

GrahamCampbell replied:
Please just replace all this with:

```php
return $pipe->{$this->method}(...$parameters);
```

ralphschindler replied:
Sounds good, I didn't realize 5.6 was the min version.  Will do.

ralphschindler committed:
Style fixes

GrahamCampbell commented on src/Illuminate/Routing/Route.php:
```diff
@@ -171,7 +171,16 @@ protected function runCallable()
-        return call_user_func_array($this->action['uses'], $parameters);
+        $callable = $this->action['uses'];
+        $parameters = array_values($parameters);
+        switch (count($parameters)) {
```
Same here. :)

GrahamCampbell commented on src/Illuminate/Routing/Route.php:
```diff
@@ -171,7 +171,16 @@ protected function runCallable()
 
-        return call_user_func_array($this->action['uses'], $parameters);
+        $callable = $this->action['uses'];
+        $parameters = array_values($parameters);
```
Note to anyone reviewing this. This will also be required to use unpacking. You cannot unpack an associative array.

ralphschindler committed:
Changes From Review:
- use php 5.6 argument unpacking (splat) in favor of switch statement
- remove one other call_user_func call previously missed

ralphschindler committed:
Style fixes

ralphschindler commented:
Should you have any additional feedback, I'll be happy to implement. Thanks for the review.

taylorotwell commented:
Thanks!

KennedyTedesco commented:
@ralphschindler Great work! Looks really good.
